### ISSUE TITLE
dapo token mean loss

### ISSUE THREAD

**Issue by yuepengs** [verl-project/verl#2776]
I'm confusing about the dapo token mean loss computation

the loss in the paper calculate the token mean within the group, while the implementation in the code compute the token mean within batch.
This difference confuses me.

<img width="1009" height="713" alt="Image" src="https://github.com/user-attachments/assets/daa4f6e0-f730-4ca1-a667-53f170b8b9b0" />

**Comment by Chandler-Bing** [comment]
+1

**Comment by vllbc** [comment]
Yes, I have also noticed this issue. Generally speaking, the number of valid tokens in each group is different, so 

$\frac{\frac{loss_{g1}}{|o_{g1}|} + \frac{loss_{g2}}{|o_{g2}|}}{2} \neq \frac{loss_{g1}+loss_{g2}}{\mid o_{g_{1}} \mid + \mid o_{g_{2}}\mid}$
(This is somewhat similar to the case where the length of the sample sequence is not equal when using gradient accumulation)

I tested it in the code and added `breakpoint()` after `loss=verl_F.masked_mean (loss_mat, loss_mask)`.

Current implementation:

<img width="1022" height="164" alt="Image" src="https://github.com/user-attachments/assets/5bfe5c4c-1071-4844-afe5-1b3c4ca737e3" />

Then it is implemented correctly according to the formula：

`n`=4，`train_batch_size`=8，so the `bs` is 32.

<img width="1142" height="612" alt="Image" src="https://github.com/user-attachments/assets/d4aad6cd-921a-4d8d-8baf-db5156e8db55" />

And the `loss_2` is:

<img width="436" height="78" alt="Image" src="https://github.com/user-attachments/assets/6bcb50f3-d970-400a-a17f-9302d871a875" />

Although they are very close, they are indeed different.

**Comment by LeonEricsson** [comment]
agreed, with a micro-batch size > 1 the implementation becomes a batch level normalization. That is fine, it avoids the question level bias, but it doesn't match the dapo formulation. See also

<img width="2048" height="868" alt="Image" src="https://github.com/user-attachments/assets/f8243b28-c095-44b7-bc9d-6de7391a9ceb" />

**Comment by tommyhuang6** [comment]
same question

**Comment by syhzcx** [comment]
@LeonEricsson   Hi! so the implementation of VERL actually calculates the token mean within each micro_batch_size_per_gpu, and if I set it to 1, it becomes a sequence mean. is my understanding correct?  Thank you!

**Comment by LeonEricsson** [comment]
> Hi! so the implementation of VERL actually calculates the token mean within each micro_batch_size_per_gpu, and if I set it to 1, it becomes a sequence mean. is my understanding correct? Thank you!

Yes, that's right. Each device performs a local forward pass on the micro batch 

https://github.com/volcengine/verl/blob/1a62568f801ba35ac1f5387e27232a2df7eac488/verl/workers/actor/dp_actor.py#L403-L439

and the subsequent length normalization 

https://github.com/volcengine/verl/blob/1a62568f801ba35ac1f5387e27232a2df7eac488/verl/trainer/ppo/core_algos.py#L887

will normalize over the entire micro-batch. If the micro-batch is only 1 response, then its to sequence normalization. 

Response groups are treated as independent samples that get shuffled and distributed across devices without preserving their grouping structure. That means we can't perform per-question / group normalization.

**Comment by syhzcx** [comment]
> > Hi! so the implementation of VERL actually calculates the token mean within each micro_batch_size_per_gpu, and if I set it to 1, it becomes a sequence mean. is my understanding correct? Thank you!
> 
> Yes, that's right. Each device performs a local forward pass on the micro batch
> 
> [verl/verl/workers/actor/dp_actor.py](https://github.com/volcengine/verl/blob/1a62568f801ba35ac1f5387e27232a2df7eac488/verl/workers/actor/dp_actor.py#L403-L439)
> 
> Lines 403 to 439 in [1a62568](/volcengine/verl/commit/1a62568f801ba35ac1f5387e27232a2df7eac488)
> 
>  for micro_batch in micro_batches: 
>      micro_batch = micro_batch.to(get_device_id()) 
>      micro_batch_metrics = {} 
>      model_inputs = {**micro_batch.batch, **micro_batch.non_tensor_batch} 
>      response_mask = model_inputs["response_mask"] 
>      old_log_prob = model_inputs["old_log_probs"] 
>      advantages = model_inputs["advantages"] 
>   
>      entropy_coeff = self.config.entropy_coeff 
>      loss_agg_mode = self.config.loss_agg_mode 
>   
>      if self.config.use_dynamic_bsz: 
>          loss_scale_factor = response_mask.shape[0] / self.config.ppo_mini_batch_size 
>      else: 
>          loss_scale_factor = 1 / self.gradient_accumulation 
>   
>      # all return: (bsz, response_length) 
>      calculate_entropy = False 
>      if entropy_coeff != 0: 
>          calculate_entropy = True 
>      entropy, log_prob = self._forward_micro_batch( 
>          model_inputs, temperature=temperature, calculate_entropy=calculate_entropy 
>      ) 
>   
>      loss_mode = self.config.policy_loss.get("loss_mode", "vanilla") 
>      # vanilla -> verl.trainer.ppo.core_algos.compute_policy_loss_vanilla 
>      # gpg -> verl.trainer.ppo.core_algos.compute_policy_loss_gpg 
>      # clip_cov -> verl.trainer.ppo.core_algos.compute_policy_loss_clip_cov 
>      policy_loss_fn = get_policy_loss_fn(loss_mode) 
>      pg_loss, pg_clipfrac, ppo_kl, pg_clipfrac_lower = policy_loss_fn( 
>          old_log_prob=old_log_prob, 
>          log_prob=log_prob, 
>          advantages=advantages, 
>          response_mask=response_mask, 
>          loss_agg_mode=loss_agg_mode, 
>          config=self.config, 
>      ) 
> and the subsequent length normalization
> 
> [verl/verl/trainer/ppo/core_algos.py](https://github.com/volcengine/verl/blob/1a62568f801ba35ac1f5387e27232a2df7eac488/verl/trainer/ppo/core_algos.py#L887)
> 
> Line 887 in [1a62568](/volcengine/verl/commit/1a62568f801ba35ac1f5387e27232a2df7eac488)
> 
>  pg_loss = agg_loss(loss_mat=pg_losses, loss_mask=response_mask, loss_agg_mode=loss_agg_mode) 
> will normalize over the entire micro-batch. If the micro-batch is only 1 response, then its to sequence normalization.
> 
> Response groups are treated as independent samples that get shuffled and distributed across devices without preserving their grouping structure. That means we can't perform per-question / group normalization.

Thanks a lot, I got it now!

**Comment by Kirrito-k423** [comment]
中文：

根据当前 `main` 分支代码和文档，这里的差异是真实存在的：verl 默认的 `actor_rollout_ref.actor.loss_agg_mode=token-mean` 是在当前 batch / micro-batch 的有效 token 上做整体平均，而不是按 DAPO 论文截图中的 prompt/group 先分别归一化再对 group 求平均。

相关代码路径是：`agg_loss(..., loss_agg_mode="token-mean")` 使用 `masked_sum(loss_mat, loss_mask) / batch_num_tokens`；policy loss 从 `config.loss_agg_mode` 读取该模式；FSDP 路径会先切成 micro-batches 再逐个 micro-batch forward/backward。文档也写明 DAPO 示例中的 `token-mean` 表示“一个批次中所有序列内所有 token 的平均值”，并且这是默认行为。

所以我对这个 issue 的理解是：当前实现更接近 batch/micro-batch token-level normalization；它可以避免按问题/序列长度带来的某些偏置，但确实不等同于 DAPO 论文中的 group-level loss formulation。若目标是严格复现论文中的 group-level normalization，需要维护者明确是否要新增/调整一个保留 response group 结构的 loss aggregation mode；现有默认行为不应被理解为严格的 group-level DAPO loss。

English:

Based on the current `main` branch code and docs, the difference reported here is real: verl's default `actor_rollout_ref.actor.loss_agg_mode=token-mean` averages over valid tokens in the current batch / micro-batch, rather than first normalizing within each prompt/group and then averaging across groups as shown in the DAPO paper formulation.

The relevant code path is: `agg_loss(..., loss_agg_mode="token-mean")` computes `masked_sum(loss_mat, loss_mask) / batch_num_tokens`; the policy loss reads this mode from `config.loss_agg_mode`; and the FSDP path prepares micro-batches before running forward/backward per micro-batch. The DAPO practice docs also describe `token-mean` as averaging policy-gradient loss over all tokens in all sequences in a batch, and note that this is the default behavior.

So my reading of this issue is: the current implementation is batch/micro-batch token-level normalization. It may avoid some question/sequence-length bias, but it is not identical to the group-level DAPO loss formulation in the paper. If strict paper-style group-level normalization is desired, maintainers would need to decide whether to add or change a loss aggregation mode that preserves response-group structure; the current default should not be interpreted as strict group-level DAPO loss.

This response was generated by a scheduled Codex GPT5.5 issue processing task.
